web/shared/src/internal_url.ts

```diff
@@ -67,3 +67,60 @@ export function by_stream_topic_url(
         maybe_get_stream_name,
     )}/topic/${encodeHashComponent(topic)}`;
 }
+/**
+ *
+ * @param url_str string containing the url
+ * @param maybe_get_stream_name function to get stream name from id
+ * @returns null if the url is malformed or belongs to a different server
+ *         [stream_id, topic] if the url is a valid narrow url. `topic` may
+ *          be undefined, if absent from the url.
+ */
+export function stream_topic_hash_from_url(
+    url_str: string,
+    maybe_get_stream_name: MaybeGetStreamName,
+): string[] {
+    let url;
+    try {
+        url = new URL(url_str);
+    } catch {
+        // url is invalid
+        return [];
+    }
+    // eslint-disable-next-line no-undef
+    if (url.origin !== window.location.origin) {
+        // not a url of the same zulip server
+        return [];
+    }
+    const paths = url.hash.split("/");
+    if (
+        paths[0] !== "#narrow" ||
+        paths[1] !== "stream" ||
+        paths[2] === undefined ||
+        paths[2].length === 0
+    ) {
+        // stream is absent from the url or is malformed
+        return [];
+    }
+    const hashed_stream = paths[2];
+    const id_hyphen = hashed_stream.indexOf("-");
+    const stream_id =
+        id_hyphen !== -1
+            ? Number.parseInt(hashed_stream.slice(0, id_hyphen), 10)
+            : Number.parseInt(hashed_stream, 10);
+    if (Number.isNaN(stream_id)) {
+        // malformed url
+        return [];
+    }
+    const stream_name = maybe_get_stream_name(stream_id);
+    if (stream_name === undefined) {
+        // stream id is invalid or untracked
+        return [];
+    }
+    const stream = decodeHashComponent(stream_name);
+    if (paths[3] !== "topic" || paths[4] === undefined || paths[4].length === 0) {
+        // topic part of the url is absent or malformed
+        return [stream];
+    }
+    const topic = decodeHashComponent(paths[4]);
+    return [stream, topic];
+}
```

Comment:
Could this make use of the techniques used in `is_same_server_message_link` to avoid so much fiddly parsing code?

(I know you can't just call that unmodified, since it looks for `near` links).

Reply:
Sure, I'll look into it.